File: zadapter3/src/main/java/com/zone/adapter3/loadmore/callback/BaseLoadMoreDelegates.java
package com.zone.adapter3.loadmore.callback;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.zone.adapter.R;
import com.zone.adapter3.loadmore.ScrollerUtils;

/**
 * [2017] by Zone
 */

public class BaseLoadMoreDelegates extends ILoadMoreDelegates {

    private View loadingView, failView, endView;

    @Override
    public void loading() {
        loadingView.setVisibility(View.VISIBLE);
        failView.setVisibility(View.GONE);
        endView.setVisibility(View.GONE);
    }

    @Override
    public void complete() {
        loadingView.setVisibility(View.GONE);
        failView.setVisibility(View.GONE);
        endView.setVisibility(View.GONE);
    }

    @Override
    public void fail() {
        loadingView.setVisibility(View.GONE);
        failView.setVisibility(View.VISIBLE);
        endView.setVisibility(View.GONE);
    }

    @Override
    public void end() {
        loadingView.setVisibility(View.GONE);
        failView.setVisibility(View.GONE);
        endView.setVisibility(View.VISIBLE);
    }

    @Override
    protected void initData(View convertView, final RecyclerView rv) {
        ViewGroup vp = (ViewGroup) convertView;
        loadingView = LayoutInflater.from(convertView.getContext())
                .inflate(R.layout.sample_common_list_footer_loading, rv, false);
        failView = LayoutInflater.from(convertView.getContext())
                .inflate(R.layout.sample_common_list_footer_network_error, rv, false);
        failView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                loading();
                ScrollerUtils.loadMore(mOnScrollRcvListener,rv);
            }
        });
        endView = LayoutInflater.from(convertView.getContext())
                .inflate(R.layout.sample_common_list_footer_end, rv, false);
        //todo  endView逻辑  最后测试
        vp.addView(loadingView);
        vp.addView(failView);
        vp.addView(endView);
    }

    @Override
    public ILoadMoreDelegates clone_() {
        return new BaseLoadMoreDelegates();
    }

    @Override
    public int getLayoutId() {
        return R.layout.vp;
    }
}

File: zadapter3/src/main/java/com/zone/adapter3kt/delegate/done/BaseLoadMoreDelegates.kt
package com.zone.adapter3kt.delegate.done

import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.zone.adapter.R
import com.zone.adapter3kt.holder.BaseHolder
import com.zone.adapter3kt.adapter.LoadMoreAdapter
import com.zone.adapter3kt.data.DataWarp
import com.zone.adapter3kt.delegate.LoadMoreViewDelegate
import com.zone.adapter3kt.loadmore.LoadingState

/**
 * [2017] by Zone
 */

class BaseLoadMoreDelegates : LoadMoreViewDelegate() {
    override fun clone_(): LoadMoreViewDelegate = BaseLoadMoreDelegates()

    override val layoutId: Int = R.layout.vp
    private var loadingView: View? = null
    private var failView: View? = null
    private var endView: View? = null

    override fun onBindViewHolder(position: Int, item: DataWarp<Any>, baseHolder: BaseHolder<androidx.recyclerview.widget.RecyclerView.ViewHolder>, payloads: List<*>) {}


    override fun loading() {
        loadingView?.visibility = View.VISIBLE
        failView?.visibility = View.GONE
        endView?.visibility = View.GONE
    }

    override fun complete() {
        loadingView?.visibility = View.GONE
        failView?.visibility = View.GONE
        endView?.visibility = View.GONE
    }

    override fun fail() {
        loadingView?.visibility = View.GONE
        failView?.visibility = View.VISIBLE
        endView?.visibility = View.GONE
    }

    override fun end() {
        loadingView?.visibility = View.GONE
        failView?.visibility = View.GONE
        endView?.visibility = View.VISIBLE
    }

    override fun initData(convertView: View, rv: ViewGroup) {
        val vp = convertView as ViewGroup
        loadingView = LayoutInflater.from(convertView.getContext())
            .inflate(R.layout.sample_common_list_footer_loading, rv, false)
        failView = LayoutInflater.from(convertView.getContext())
            .inflate(R.layout.sample_common_list_footer_network_error, rv, false)
        failView?.setOnClickListener {
            if (rv is androidx.recyclerview.widget.RecyclerView && rv.adapter is LoadMoreAdapter<*>) {
                (rv.adapter as LoadMoreAdapter<*>).loadOnScrollListener?.loadMore(rv)
            }
        }
        endView = LayoutInflater.from(convertView.getContext())
            .inflate(R.layout.sample_common_list_footer_end, rv, false)
        vp.addView(loadingView)
        vp.addView(failView)
        vp.addView(endView)
        if (adapter is LoadMoreAdapter) {
            when ((adapter as LoadMoreAdapter<*>).loadOnScrollListener!!.loadingStateInner) {
                LoadingState.NO_SHOW -> {
                }
                LoadingState.LOADING -> loading()
                LoadingState.END -> end()
                LoadingState.FAIL -> fail()
                LoadingState.COMPLETE -> complete()
            }
        }

    }
}
